halo/evmengine/keeper/abci_internal_test.go

```diff
@@ -0,0 +1,544 @@
+var _ etypes.CPayloadProvider = (*MockCPayloadProvider)(nil)
+
+type MockEngineAPI struct {
+	fuzzer                  *fuzz.Fuzzer
+	Mock                    ethclient.EngineClient // avoid repeating the implementation but also allow for custom implementations of mocks
+	BlockNumberFunc         func(ctx context.Context) (uint64, error)
+	BlockByNumberFunc       func(ctx context.Context, number *big.Int) (*types.Block, error)
+	ForkchoiceUpdatedV2Func func(ctx context.Context, update eengine.ForkchoiceStateV1,
+		payloadAttributes *eengine.PayloadAttributes) (eengine.ForkChoiceResponse, error)
+}
+
+func (m *MockEngineAPI) ChainID(ctx context.Context) (*big.Int, error) {
+	// TODO implement me
+	panic("implement me")
+}
+
+func (m *MockEngineAPI) BlockByHash(ctx context.Context, hash common.Hash) (*types.Block, error) {
+	// TODO implement me
+	panic("implement me")
+}
+
+func (m *MockEngineAPI) HeaderByHash(ctx context.Context, hash common.Hash) (*types.Header, error) {
+	// TODO implement me
+	panic("implement me")
+}
+
+func (m *MockEngineAPI) HeaderByNumber(ctx context.Context, number *big.Int) (*types.Header, error) {
+	// TODO implement me
+	panic("implement me")
+}
+
+func (m *MockEngineAPI) TransactionCount(ctx context.Context, blockHash common.Hash) (uint, error) {
+	// TODO implement me
+	panic("implement me")
+}
+
+func (m *MockEngineAPI) TransactionInBlock(ctx context.Context, blockHash common.Hash, index uint) (*types.Transaction, error) {
+	// TODO implement me
+	panic("implement me")
+}
+
+func (m *MockEngineAPI) SubscribeNewHead(ctx context.Context, ch chan<- *types.Header) (ethereum.Subscription, error) {
+	// TODO implement me
+	panic("implement me")
+}
+
+func (m *MockEngineAPI) BalanceAt(ctx context.Context, account common.Address, blockNumber *big.Int) (*big.Int, error) {
+	// TODO implement me
+	panic("implement me")
+}
+
+func (m *MockEngineAPI) StorageAt(ctx context.Context, account common.Address, key common.Hash, blockNumber *big.Int) ([]byte, error) {
+	// TODO implement me
+	panic("implement me")
+}
+
+func (m *MockEngineAPI) CodeAt(ctx context.Context, account common.Address, blockNumber *big.Int) ([]byte, error) {
+	// TODO implement me
+	panic("implement me")
+}
+
+func (m *MockEngineAPI) NonceAt(ctx context.Context, account common.Address, blockNumber *big.Int) (uint64, error) {
+	// TODO implement me
+	panic("implement me")
+}
+
+func (m *MockEngineAPI) SyncProgress(ctx context.Context) (*ethereum.SyncProgress, error) {
+	// TODO implement me
+	panic("implement me")
+}
+
+func (m *MockEngineAPI) CallContract(ctx context.Context, call ethereum.CallMsg, blockNumber *big.Int) ([]byte, error) {
+	// TODO implement me
+	panic("implement me")
+}
+
+func (m *MockEngineAPI) EstimateGas(ctx context.Context, call ethereum.CallMsg) (uint64, error) {
+	// TODO implement me
+	panic("implement me")
+}
+
+func (m *MockEngineAPI) SuggestGasPrice(ctx context.Context) (*big.Int, error) {
```

Comment:
can remove implementme functions, just embed the interface you want to implement in the struct itself